[exploreV2] mapStateToProps for fields

mistercrunch opened the pull request:
This is part of the effort to break down https://github.com/airbnb/superset/pull/1788 and builds up on https://github.com/airbnb/superset/pull/1868 (landed).

* introducing a `mapStateToProps` to field configuration
* making the backend endpoint `fetch_datasource_metadata` do just that, no more backend logic that "broadcasts" field options into the state, instead [more generically] the fields define how they pick from the state to define their props (often `choices` in the current use cases)

Note:
`fetch_datasource_metadata` could be simplified further to expose the raw array of objects of columns and metrics, and more logic to build options could be moved to the js side. It's judged out of scope for this PR.

mistercrunch committed:
Controls support for mapStateToProps

bkyryliuk commented on superset/assets/javascripts/explorev2/components/ControlPanelsContainer.jsx:
```diff
@@ -17,48 +18,58 @@ const propTypes = {
+    this.props.actions.setFieldValue(this.props.datasource_type, name, value);
+  }
+  getFieldData(fs) {
+    const fieldOverrides = this.fieldOverrides.bind(this)();
```
I am curious, why can't u just use `this.fieldOverrides()` instead of `this.fieldOverrides.bind(this)()`?

mistercrunch replied:
ECMAScript 6 `class` declarations don't bind the method to `this`. I should just bind the functions in the constructor to make this cleaner.

bkyryliuk commented on superset/models.py:
```diff
@@ -655,6 +655,34 @@ def explore_url(self):
+            'all_cols': all_cols,
+            'filterable_cols': self.filterable_column_names,
+        }
+        if (self.type == 'table'):
```
what about druid grains?

mistercrunch replied:
I think they're static so they are hard coded in the fields configuration

bkyryliuk commented:
LGTM

mistercrunch committed:
Binding methods in the constructor

ascott commented on superset/assets/javascripts/explorev2/components/ControlPanelsContainer.jsx:
```diff
@@ -17,48 +18,65 @@ const propTypes = {
 };
 
 class ControlPanelsContainer extends React.Component {
+  constructor() {
```
i believe the constructor takes props in `constructor(props)`, and that `super` should be called with those props `super(props);`

mistercrunch replied:
oh right!

ascott commented on superset/assets/javascripts/explorev2/stores/fields.js:
```diff
@@ -552,42 +556,52 @@ export const fields = {
     default: null,
     description: 'This define the element to be plotted on the chart',
+    mapStateToProps: (state) => ({
+      choices: state.datasource.gb_cols,
```
looks like we should check for datasource existence here too
`(state.datasource) ? state.datasource.gb_cols : [],`

mistercrunch replied:
totally, I'm not sure how I missed that in my copy-pasta...

ascott commented on superset/assets/javascripts/explorev2/stores/fields.js:
```diff
@@ -961,10 +970,12 @@ export const fields = {
                  'Either a numerical column or `Auto`, which scales the point based ' +
                  'on the largest cluster',
+    mapStateToProps: (state) => ({
+      choices: state.fields.point_radius.choices,
```
will `state.fields.point_radius` even be null?

ascott commented on superset/assets/javascripts/explorev2/stores/fields.js:
```diff
@@ -993,15 +1004,15 @@ export const fields = {
 
   viewport_latitude: {
-    type: 'IntegerField',
+    type: 'TextField',
```
one reason i was using a type called `IntegerField` was so we could do number validation on the field in that component, but we could handle this in a different way in the future.

mistercrunch replied:
I figured that was the intent and had some thoughts about how we may want to express that in the future. For now I wanted to avoid the "virtual fields" that had no direct corresponding components.

I'm thinking that controls configuration should allow for a `validators` prop in the form of an array of functions. We can define a set of common validators (validators.integer, validators.non_null, validators.at_least_one), process them in the FieldSet controller, and give cues to the users (maybe a red exclamation point).

The validator function would receive the `value`, the field configuration, and be expected to return a comprehensive message. Complex fields will be allowed to define complex validators if needed.

mistercrunch replied:
Create an issue with the information above:
https://github.com/airbnb/superset/issues/1912

ascott replied:
sweet! thanks for documenting that.

ascott commented on superset/assets/javascripts/explorev2/components/ControlPanelsContainer.jsx:
```diff
@@ -119,6 +141,7 @@ function mapStateToProps(state) {
     alert: state.controlPanelAlert,
     isDatasourceMetaLoading: state.isDatasourceMetaLoading,
     fields: state.fields,
+    exploreState: state,
```
do we need to pass the entire state object in here? could we pass just the keys we need?

ascott replied:
looks like we only need `state.datasource`...

mistercrunch replied:
related to my previous comment in this PR

ascott commented on superset/assets/javascripts/explorev2/stores/fields.js:
```diff
@@ -31,13 +31,15 @@ export const fields = {
     clearable: false,
     default: null,
-    choices: [],
+    mapStateToProps: (state) => ({
```
this name seems a bit confusing to me since we use the same function name in components to map state from the store to props. could we call this `getChoices`?

mistercrunch replied:
related to my previous comment in this PR

ascott replied:
gotcha, i see what you are going for now.

ascott commented on superset/assets/javascripts/explorev2/components/ControlPanelsContainer.jsx:
```diff
@@ -17,48 +18,65 @@ const propTypes = {
+      const overrideData = fieldOverrides[fs];
+      fieldData = Object.assign({}, fieldData, overrideData);
+    }
+    if (fieldData.mapStateToProps) {
```
i think this could be more clear as the following:
- renaming `mapStateToProps` to `getChoices` and only passing in the datasource, rather than the entire state.

```js
if (fieldData.getChoices) {
  Object.assign(fieldData, fieldData.getChoices(this.props.datasource));
}
```

mistercrunch replied:
The premise of this PR is to have a very flexible framework where controls define if/what props should come from the state, and have full access to the state. If we want to evolve the state to have database configuration information, or information about user preferences, controls should be able to `mapStateToProps` using those.

ascott commented on superset/assets/spec/javascripts/explorev2/components/FieldSetRow_spec.js:
```diff
@@ -25,7 +20,6 @@ describe('FieldSetRow', () => {
 
   it('renders a FieldSet for each item in fieldSets array', () => {
-    const length = defaultProps.fieldSets.length;
-    expect(wrapper.find(FieldSet)).to.have.lengthOf(length);
```
why don't we want to assert how many `FieldSet` components are rendered anymore? i'm not really sure how this change is testing the `FieldSetRow` component.

mistercrunch replied:
The new component is more generic than the previous one as I moved much of the logic in the parent component. Now it really just makes a row out of an array of arbitrary nodes. I'll fix up the naming and add a few more test cases.

mistercrunch committed:
Adressing comments

mistercrunch commented:
Thanks for the review @bkyryliuk & @ascott , I addressed the comments and things I did not make clear around around the design of `mapStateToProps` and how to handle validators in the future.

Let me know what you think!

bkyryliuk commented:
still looks good to me. It would be great to have some comments / docs showing examples how datasource data is mapped to the controls, but that can be done later on.

ascott commented:
thanks for the changes, LGTM!

mistercrunch committed:
Fixing tests
